src/transport/http2.rs

```diff
@@ -201,19 +167,104 @@ impl<In: RpcMessage, Out: RpcMessage> ServerChannel<In, Out> {
+    async fn handle_one_http2_request(
+        req: Request<Body>,
+        accept_tx: Sender<(Receiver<In>, Sender<Out>)>,
+    ) -> Result<Response<Body>, String> {
+        let (req_tx, req_rx) = flume::bounded::<In>(32);
+        let (res_tx, res_rx) = flume::bounded::<Out>(32);
+        accept_tx
+            .send_async((req_rx, res_tx))
+            .await
+            .map_err(|_e| "unable to send")?;
+
+        Self::spawn_recv_forwarder(req, req_tx);
+        let body = res_rx.into_stream().map(|out| {
+            let data = bincode::serialize(&out)?;
+            // TODO: check size
+            Ok::<Vec<u8>, bincode::Error>(data)
+        });
+
+        // Create a response with the response body channel as the response body
+        let response = Response::builder()
+            .status(StatusCode::OK)
+            .body(Body::wrap_stream(body))
+            .map_err(|_| "unable to set body")?;
+        Ok(response)
+    }
+
+    /// Spawns a task which forwards requests from the network to a flume channel.
+    ///
+    /// This task will read frames from the network, decodes them using bincode and forward them
+    /// to the flume channel used to receive a request stream.
+    ///
+    /// Could also have been implemented as a stream map.
+    ///
+    /// If there is a network error, decoding failure, the flume channel closes or the request
+    /// stream is simply ended this task will terminate.  So it is fine to ignore the returned
+    /// [`JoinHandle`].
+    ///
+    /// The HTTP2 request comes from *req* and the decoded version is sent to `req_tx`.
+    fn spawn_recv_forwarder(req: Request<Body>, req_tx: Sender<In>) -> JoinHandle<()> {
+        tokio::spawn(async move {
+            let mut stream = req.into_body();
+
+            // This assumes each chunk received corresponds to a single HTTP2 frame.
+            while let Some(chunk) = stream.next().await {
+                match chunk {
+                    Ok(chunk) => match bincode::deserialize::<In>(chunk.as_ref()) {
+                        Ok(msg) => {
+                            log!("server got msg {msg:?}");
+                            match req_tx.send_async(msg).await {
+                                Ok(()) => {}
+                                Err(_cause) => {
+                                    // channel closed
+                                    break;
+                                }
+                            }
+                        }
+                        Err(_cause) => {
+                            // deserialize error
+                            break;
+                        }
+                    },
+                    Err(_cause) => {
+                        // error from the network layer
+                        break;
+                    }
+                }
+            }
+        })
     }
 }
 
+// This does not want or need RpcMessage to be clone but still want to clone the
```

Comment:
I was thinking about just adding RpcMessage: Clone to make the derive macros happy, but I found that they sometimes get confused even then.

E.g. RpcMessage requires Debug, but nevertheless I found that sometimes I had to manually impl Debug. Or maybe I imagined this.

Reply:
Yeah, it doesn't seem terrible to require a message to be clone.  But also it doesn't have to be, so I''m pretty happy with this tradeoff as it is.  Just wanted to have this explained since it's not always obvious if you haven't encountered this before.